engine/new_runtime/src/memories/noncontiguous_memory.rs

```diff
@@ -0,0 +1,372 @@
+    }
+}
+
+
+impl<P: BoxProvider> NonContiguousMemory<P>
+{
+    // Check that the given configurations are correct for allocation
+    fn check_config(config: &LockedConfiguration<P>) -> Result<(), MemoryError> {
+        match config {
+            NCRamAndFileConfig(Some(size)) => {
+                if *size != NC_DATA_SIZE {
+                    Err(NCSizeNotAllowed)
+                } else {
+                    Ok(())
+                }
+            },
+            NCRamConfig(Some(size)) => {
+                if *size != NC_DATA_SIZE  {
+                    Err(NCSizeNotAllowed)
+                } else {
+                    Ok(())
+                }
+            },
+            _ => Err(ConfigurationNotAllowed)
+        }
+    }
+
+    fn get_buffer_from_shard1(&self) -> &Buffer<u8> {
+        match &self.shard1 {
+            RamShard(buf) => buf,
+            _ => panic!("{}", IMPOSSIBLE_CASE),
+        }
+    }
+
+
+    // Refresh the shards to increase security, may be called every _n_ seconds or
+    // punctually
+    fn refresh(self) -> Result<Self, MemoryError> {
+        NonContiguousMemory::check_config(&self.config)?;
+        let random = P::random_vec(NC_DATA_SIZE).expect("Failed to generate random vec");
+        let mut hash_of_random = [0u8; NC_DATA_SIZE];
+
+        // Refresh shard1
+        let data_of_old_shard1: &[u8] = &self.get_buffer_from_shard1().borrow();
+        let new_data1 = xor(data_of_old_shard1, &random);
+        let new_shard1 = RamShard(Buffer::alloc(&new_data1, BufferConfig(NC_DATA_SIZE))?);
+
+        let new_shard2;
+        let new_config;
+        let mut hash_of_old_shard1 = [0u8; NC_DATA_SIZE];
+        let mut hash_of_new_shard1 = [0u8; NC_DATA_SIZE];
+        sha::SHA256(&data_of_old_shard1, &mut hash_of_old_shard1);
+        sha::SHA256(&new_data1, &mut hash_of_new_shard1);
+
+        match &self.shard2 {
+            RamShard(b2) => {
+                let new_data2 = xor(&b2.borrow(), &hash_of_old_shard1);
+                let new_data2 = xor(&new_data2, &hash_of_new_shard1);
+                new_shard2 = RamShard(Buffer::alloc(&new_data2, BufferConfig(NC_DATA_SIZE))?);
+                new_config = NCRamConfig(Some(NC_DATA_SIZE));
+            },
+            FileShard(fm) => {
+                let buf = fm.unlock(FileConfig(None))
+                    .expect("Failed to unlock file memory");
+                let new_data2 = xor(&buf.borrow(), &hash_of_old_shard1);
+                let new_data2 = xor(&new_data2, &hash_of_new_shard1);
+                let new_fm = FileMemory::alloc(&new_data2,
+                                               FileConfig(Some(NC_DATA_SIZE)))?;
+                new_shard2 = FileShard(new_fm);
+                // new_shard2 = FileShard(fm.lock(new_data2, FileConfig(Some(NC_DATA_SIZE)))?);
+                new_config = NCRamAndFileConfig(Some(NC_DATA_SIZE));
+            },
+            _ => panic!("{}", IMPOSSIBLE_CASE)
```

Comment:
Isn't it better to panic if we reach a case that should never happen and is present just because Rust requires exhaustive pattern matching?

Reply:
How about using the rust idiomatic `unreachable!("Config was already checked.")`? Does the same as `panic`, but emphasizes that this case will **never** happen because we already checked it above.

Reply:
Sounds perfect for this use case, thanks

Reply:
Depends on what we want to achieve. If we extend `LockedConfiguration`, but do not provide a matching implementation for new types, I would expect an error. Since `refresh` doesn't accept a config type from the outside, this is a logical follow up error. I would also suggest to avoid `unreachable!()` in that case, because this case should be handled earlier. I do understand the desired implication. Maybe `unreachable!()` can be used earlier, if we intent to "panic with reason" on this crucial part.

Reply:
@felsweg-iota before this match block, in the same method, it is already checked if the config is one of the two valid ones, and otherwise an error is returned:
https://github.com/iotaledger/stronghold.rs/blob/43cde6b3be50a4f0b433e4551b3d17648e65f617/engine/new_runtime/src/memories/noncontiguous_memory.rs#L132
So it is already doing exactly what you are proposing.

Reply:
@elenaf9 no. I see no reason to introduce `unreachable!()` here, where some returned error could indicate a failure to allocate memory. I would contain `panic` etc. to more sensitive regions. While this code indicates prior-allocation, but config checking, an error would be appropriate.

Reply:
No actually, you are right @felsweg-iota. I did not read through the code carefully enough and missed that it is matching `self.shard2` here, sorry. I thought it was matching on the `config` parameter.
In this case I agree with you that we should return an error. It could be the case that e.g. zeroize  was called, in which case it would cause a panic. Sorry for the confusion.